ledger/complete/mtrie/node/node.go

```diff
@@ -122,12 +122,98 @@ func NewInterimNode(height int, lchild, rchild *Node) *Node {
 	return n
 }
 
+// deepCopy
+func (n *Node) deepCopy() *Node {
+	var h hash.Hash
+	copy(h[:], n.hashValue[:])
+
+	var p ledger.Path
+	copy(p[:], n.path[:])
+
+	var lChildCopy, rChildCopy *Node
+	if n.lChild != nil {
+		lChildCopy = n.lChild.deepCopy()
+	}
+	if n.rChild != nil {
+		rChildCopy = n.rChild.deepCopy()
+	}
+
+	return &Node{
+		lChild:    lChildCopy,
+		rChild:    rChildCopy,
+		height:    n.height,
+		path:      p,
+		payload:   n.payload.DeepCopy(),
+		maxDepth:  n.maxDepth,
+		regCount:  n.regCount,
+		hashValue: h,
+	}
+}
+
 // computeAndStoreHash computes the node's hash value and
 // stores the result in the nodes internal `hashValue` field
 func (n *Node) computeAndStoreHash() {
 	n.hashValue = n.computeHash()
 }
 
+// Compactify an expanded leaf into its most concise representation. The
+// compactified representation of a default node is `nil`. For a node that only has a
+// _single_ child that is itself a leaf, this method returns a new, fully compactified leaf.
+// Returns:
+//  * (n, false) if the node cannot be compactified, we return the original node `n`
+//  * (cn, true) if the node can be compactified, where cn is a newly created compactified leaf
+func (n *Node) Compactify() (*Node, bool) {
+	if n.IsLeaf() {
+		return n, false
+	}
+
+	// if leaf return it as is
+	// if non leaf bubble up
+	lChildEmpty := true
+	rChildEmpty := true
+	if n.lChild != nil {
+		lChildEmpty = n.lChild.IsDefaultNode()
```

Comment:
If we add the check `if n.IsADefaultNode()` on the top as suggested by Alex, we shouldn't need to make these checks anymore : since we call `compactify` in a bottom up manner, both children must have been already compactified, so the only check we need here is for `n.lChild == nil` and `n.rChild == nil`

Reply:
Update: it looks like `compactify` is only called on interim nodes, and not on leaves 🤔  so I think the check you wrote is needed.